src/lib/coda_base/user_command_memo.ml

```diff
@@ -1,21 +1,99 @@
 open Core
-include Random_oracle.Digest
+open Crypto_params
 
-(*TODO Currently sha-ing a string. Actual data in the memo needs to be decided *)
-let max_size_in_bytes = 1000
+module Stable = struct
+  module V1 = struct
+    module T = struct
+      open Base58_check
 
-let create_exn s =
-  let max_size = 1000 in
-  if Int.(String.length s > max_size_in_bytes) then
-    failwithf !"Memo data too long. Max size = %d" max_size ()
-  else Random_oracle.digest_string s
+      type t = string
+      [@@deriving bin_io, sexp, eq, compare, hash, version {unnumbered}]
 
-let dummy = create_exn ""
+      let version_byte = Version_bytes.user_command_memo
 
-include Codable.Make_of_string (struct
-  type nonrec t = t
+      let to_string (memo : t) : string = encode ~version_byte ~payload:memo
 
-  let to_string (memo : t) = Base64.encode_string (memo :> string)
+      let of_string (s : string) : t = decode_exn ~version_byte s
+    end
 
-  let of_string = Fn.compose of_string Base64.decode_exn
-end)
+    include T
+    include Codable.Make_of_string (T)
+  end
+
+  module Latest = V1
+end
+
+type t = Stable.Latest.t [@@deriving sexp, eq, compare, hash]
+
+[%%define_locally
+Stable.Latest.(to_yojson, of_yojson, to_string, of_string)]
+
+exception Invalid_user_memo_length
+
+exception Too_long_digestible_string
+
+let max_digestible_string_length = 1000
+
+[%%define_locally
+Random_oracle.Digest.(length_in_bytes, length_in_triples)]
+
+let create_by_digesting_string_exn s =
+  if Int.(String.length s > max_digestible_string_length) then
+    raise Too_long_digestible_string ;
+  (Random_oracle.digest_string s :> t)
+
+let create_from_bytes32_exn bytes : t =
+  if not (Int.equal (Bytes.length bytes) length_in_bytes) then
+    raise Invalid_user_memo_length ;
+  Bytes.to_string bytes
+
+let create_from_string32_exn s : t =
```

Comment:
Can you make versions of these that are safe as well? I think the graphql stuff should use the non exn versions too.